lib/spack/spack/concretize.py

```diff
@@ -218,9 +218,11 @@ def concretize_architecture(self, spec):
         sys_arch = spack.spec.ArchSpec(spack.architecture.sys_type())
         spec_changed = False
 
-        if spec.architecture is None:
+        if spec.architecture is None and root_arch is None:
```

Comment:
Instead of root_arch, I think this should use an idiom similar to `concretize_compiler` and search up the DAG for the nearest architecture that is set. Suppose foo depends on bar depends on baz. Then if a user specifies `foo os=be ^bar os=fe`, the baz in that spec should concretize to the front-end architecture, not the back-end. @mamelara have you checked that case yet?

Maybe it should concretize the three portions separately, but that would be a bigger change.